stumpy/stump_tiles.py

```diff
@@ -0,0 +1,643 @@
+        Set to `True` if this is a self-join. Otherwise, for AB-join, set this to
+        `False`. Default is `True`.
+
+    Returns
+    -------
+    None
+
+    Notes
+    -----
+    `DOI: 10.1007/s10115-017-1138-x \
+    <https://www.cs.ucr.edu/~eamonn/ten_quadrillion.pdf>`__
+
+    See Section 4.5
+
+    The above reference outlines a general approach for traversing the distance
+    matrix in a diagonal fashion rather than in a row-wise fashion.
+
+    `DOI: 10.1145/3357223.3362721 \
+    <https://www.cs.ucr.edu/~eamonn/public/GPU_Matrix_profile_VLDB_30DraftOnly.pdf>`__
+
+    See Section 3.1 and Section 3.3
+
+    The above reference outlines the use of the Pearson correlation via Welford's
+    centered sum-of-products along each diagonal of the distance matrix in place of the
+    sliding window dot product found in the original STOMP method.
+    """
+    m_inverse = 1.0 / m
+    constant = (m - 1) * m_inverse * m_inverse  # (m - 1)/(m * m)
+    chunk_size = config.STUMPY_DIAGONALS_CHUNK_SIZE
+
+    for tile_idx in tiles_ranges[thread_idx]:
+        # break if tile index is negative
+        if tile_idx < 0:
+            break
+
+        y_offset, x_offset = tiles[tile_idx][0]
+        tile_height, tile_width = tiles[tile_idx][1]
+        diags = np.arange(tiles[tile_idx][2][0], tiles[tile_idx][2][1])
+
+        # continue if tile is indicated invalid
+        if tiles[tile_idx][3][1] == 0:
+            continue
+
+        for chunk_idx in range(0, diags.shape[0], chunk_size):
+            # get current chunk size, i.e. number of diagonals to traverse together
+            current_chunk_size = min(chunk_size, diags.shape[0] - chunk_idx)
+            # get diagonals to traverse together
+            diags_chunk = diags[chunk_idx : chunk_idx + current_chunk_size]
+
+            iter_ranges = np.zeros((current_chunk_size, 3), dtype=np.uint64)
+            covariances = np.zeros(current_chunk_size, dtype=np.float64)
+
+            longest_diag_len = -1
+            for diag_idx in range(diags_chunk.shape[0]):
+                diag = diags_chunk[diag_idx]
+
+                # store index of diagonal
+                iter_ranges[diag_idx][0] = diag - y_offset + x_offset
+
+                # store range of diagonal to traverse
+                if diag >= 0:
+                    iter_ranges[diag_idx][1] = y_offset
+                    iter_ranges[diag_idx][2] = (
+                        min(tile_height, tile_width - diag) + y_offset
+                    )
+                else:
+                    iter_ranges[diag_idx][1] = y_offset - diag
+                    iter_ranges[diag_idx][2] = (
+                        min(tile_height, tile_width - diag) + y_offset
+                    )
+
+                # record length of longest diagonal
+                longest_diag_len = max(
+                    longest_diag_len,
+                    iter_ranges[diag_idx][2] - iter_ranges[diag_idx][1],
+                )
+
+            m = np.uint64(m)
```

Comment:
You shouldn't need to cast this as `m` is already an int

Reply:
So I'm casting to `uint64` here, so it doesn't silently convert indices to `float64`

Reply:
If that's the case, then you should move this outside of all of the looping. Otherwise, you are casting more than once.